File: marvelHeroes/src/main/java/com/futureworkshops/marvelheroes/domain/repositories/character/MarvelCharacterRepository.java
/*
 * Copyright (c) 2018 FutureWorkshops. All rights reserved.
 */

package com.futureworkshops.marvelheroes.domain.repositories.character;

import android.support.annotation.NonNull;
import android.util.Log;

import com.futureworkshops.marvelheroes.data.network.RestManager;
import com.futureworkshops.marvelheroes.data.network.dto.CharacterDto;
import com.futureworkshops.marvelheroes.domain.model.Character;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import io.reactivex.Single;

/**
 * Created by stelian on 03/04/2018.
 */

public class MarvelCharacterRepository {
    
    public static final int AVENGERS__SERIES_ID = 3621;
    
    private RestManager restManager;
    
    class Token {
        String token;
    }
    
    @Inject
    public MarvelCharacterRepository(RestManager restManager) {
        this.restManager = restManager;
    }
    
    public Single<List<Character>> getAvengersCharacters() {
        CharacterQuery filter = new CharacterQuery.Builder()
            .serie(AVENGERS__SERIES_ID)
            .orderByName(true)
            .limit(50)
            .build();
        
        return getCharactersWithQuery(filter);
        
    }
    
    public Single<List<Character>> getCharacterDetails(int characterId) {
        return restManager.getCharacterDetails(String.valueOf(characterId))
            .flatMap(response -> Single.just(mapDtoToModel(response)));
    }
    
    public Single<List<Character>> getAllCharacters(int offset, int limit) {
        CharacterQuery filter = new CharacterQuery.Builder()
            .offset(offset)
            .limit(limit)
            .build();
        
        
        
        
        return getCharactersWithQuery(filter);
    }
    
    
    private Single<List<Character>> getCharactersWithQuery(@NonNull CharacterQuery characterQuery) {
        return restManager.getCharactersWithQuery(characterQuery.toMap())
            .flatMap(response -> Single.just(mapDtoToModel(response)));
    }
    
    @NonNull
    private List<Character> mapDtoToModel(List<CharacterDto> characterDtos) {
        final List<Character> characters = new ArrayList<>();
        
        for (CharacterDto dto : characterDtos) {
            characters.add(CharacterMapper.dtoToCharacter(dto));
        }
        return characters;
    }
    
}

File: marvelHeroes/src/main/java/com/futureworkshops/marvelheroes/domain/repositories/character/MarvelCharacterRepository.kt
/*
 * Copyright (c) 2018 FutureWorkshops. All rights reserved.
 */

package com.futureworkshops.marvelheroes.domain.repositories.character

import com.futureworkshops.marvelheroes.data.network.RestManager
import com.futureworkshops.marvelheroes.data.network.dto.MarvelCharacterDto
import com.futureworkshops.marvelheroes.domain.model.MarvelCharacter
import io.reactivex.Single
import java.util.*
import javax.inject.Inject

/**
 * Created by stelian on 03/04/2018.
 */

class MarvelCharacterRepository @Inject
constructor(private val restManager: RestManager) {
    
    val avengersCharacters: Single<List<MarvelCharacter>>
        get() {
            val filter = CharacterQuery.Builder()
                    .serie(AVENGERS__SERIES_ID)
                    .orderByName(true)
                    .limit(50)
                    .build()
            
            return getCharactersWithQuery(filter)
            
        }
    
    internal inner class Token {
        var token: String? = null
    }
    
    fun getCharacterDetails(characterId: Int): Single<List<MarvelCharacter>> {
        return restManager.getCharacterDetails(characterId.toString())
                .flatMap { response -> Single.just(mapDtoToModel(response)) }
    }
    
    fun getAllCharacters(offset: Int, limit: Int): Single<List<MarvelCharacter>> {
        val filter = CharacterQuery.Builder()
                .offset(offset)
                .limit(limit)
                .build()
        return getCharactersWithQuery(filter)
    }
    
    
    private fun getCharactersWithQuery(characterQuery: CharacterQuery): Single<List<MarvelCharacter>> {
        return restManager.getCharactersWithQuery(characterQuery.toMap())
                .flatMap { response -> Single.just(mapDtoToModel(response)) }
    }
    
    private fun mapDtoToModel(characterDtos: List<MarvelCharacterDto>): List<MarvelCharacter> {
        val characters = ArrayList<MarvelCharacter>()
        
        for (dto in characterDtos) {
            characters.add(CharacterMapper.dtoToCharacter(dto))
        }
        return characters
    }
    
    companion object {
        const val AVENGERS__SERIES_ID = 3621
    }
    
}
